### PR TITLE
Reverse HOTP

### PR DESCRIPTION
**Pull request by szszszsz**
Add the reverse HOTP feature

- CLI: https://github.com/Nitrokey/pynitrokey/pull/280
- Tests: https://github.com/Nitrokey/pynitrokey/pull/280/files#diff-8fb21877f9a0cc6c88d7d4076e5e6ab211a2bb69dd9d7cfa795c3297e1f4e9e7

## Tasks
- [x] return error on the counter overflow
- [x] discuss, if on-credential operations (get code, verify code, or both) should be limited during its registration
- [x] integrate with embedded runner
- [ ] support LED control - done partially, see https://github.com/Nitrokey/oath-authenticator/issues/4
- [x] support touch button
- [ ] extension: PIN verification as an alternative to the VALIDATE command, https://github.com/Nitrokey/oath-authenticator/issues/6
- [ ] optional: finish Trussed Virt-based example and tests , https://github.com/Nitrokey/oath-authenticator/issues/7
- [x] discuss, if the error codes should be more specific (its basically success or error now)
- [x] use u64 as the input for the `verify_code`, instead of the currently string

## Description

Verify the HOTP code coming from a PC host, and show visually to user,    that the code is correct or not, with a green or red LED respectively.     Does not need authorization by design.
    
See more at:
 -   https://github.com/Nitrokey/nitrokey-hotp-verification#verifying-hotp-code
  
  Solution contains means to avoid desynchronization between the host's and device's counters. Device calculates up to 9 values ahead of its current counter to find the matching code (in total it calculates HOTP code for 10 subsequent counter positions). In case:
    
- no code would match - the on-device counter will not be changed;
- incoming code parsing would fail - the on-device counter will not be changed;
- code would match, but with some counter's offset (up to 9) - the on-device counter will be set to matched code-generated HOTP counter and incremented by 1;
- code would match, and the code matches counter without offset - the counter will be incremented by 1.
    
Device will stop verifying the HOTP codes in case, when the difference between the host and on-device counters will be greater or equal to 10.

cc @sosthene-nitrokey

### CODE REVIEW

**Comment by szszszsz** [comment]
Let's consider this being ready to review. The remaining tasks' results will be supplied in separate PRs.

**Comment by szszszsz** [comment]
Note: confirmed to work in the NK3's embedded runner (`embedded-runner-lpc55`):
- https://github.com/Nitrokey/nitrokey-3-firmware/compare/embedded-runner-lpc55...otp-app-ctaphid
- https://github.com/Nitrokey/nitrokey-3-firmware/pull/97

File: src/authenticator.rs [hunk 1]
```diff
@@ -813,4 +911,48 @@ impl<T> hid::App for Authenticator<T>
...
+/// ```
+pub fn convert_string_to_integer(str: &[u8]) -> iso7816::Result<u32> {
+    if str.len() > 8 {
+        // Handle small codes only for now
+        return Err(Status::ConditionsOfUseNotSatisfied);
+    }
+    let mut number = 0u32;
+    for e in str.iter().rev().enumerate() {
+        let (i, digit) = e;
+        if digit < &('0' as u8) || digit > &('9' as u8) {
```
Comment (sosthene-nitrokey) [inline src/authenticator.rs:945]: You can probably use the standard library for the test and conversion:

```rust
let Some(digit_value) = (digit as char).to_digit(10) else {
    return Err(…);
};
```

File: src/command.rs [hunk 2]
```diff
@@ -121,6 +123,30 @@ impl<'l, const C: usize> TryFrom<&'l Data<C>> for Validate<'l> {
...
+        let first: TaggedSlice = decoder.decode().unwrap();
+        assert!(first.tag() == (oath::Tag::Name as u8).try_into().unwrap());
+        let label = first.as_bytes();
+
+        let slice: TaggedSlice = decoder.decode().unwrap();
+        assert!(slice.tag() == (oath::Tag::Response as u8).try_into().unwrap());
+        let response = slice.as_bytes();
+
+        Ok(VerifyCode { label, response })
+    }
```
Comment (sosthene-nitrokey) [inline src/command.rs:147]: This function should return errors rather than unwrapping and asserting

File: src/authenticator.rs [hunk 3]
```diff
@@ -698,6 +689,114 @@ where
...
+            let code = self.calculate_hotp_code_for_counter(credential, counter);
+            if code == code_in {
+                found = Some(counter);
+                break;
+            }
+        }
+
+        if found.is_none() {
+            // Failed verification
+            // TODO blink red LED infinite times, highest priority
```
Comment (sosthene-nitrokey) [inline src/authenticator.rs:749]: This TODO seems important.

This is blocked by trussed only offering and "Wink" without more configuration right?

File: src/authenticator.rs [hunk 4]
```diff
@@ -698,6 +689,114 @@ where
...
+            return Err(Status::VerificationFailed);
+        }
+
+        // TODO blink green LED for 10 seconds, highest priority
```
Comment (sosthene-nitrokey) [inline src/authenticator.rs:753]: This TODO seems important

**Review by sosthene-nitrokey — CHANGES_REQUESTED** [review]

File: src/authenticator.rs [hunk 5]
```diff
@@ -698,6 +689,114 @@ where
...
+    /// Solution contains a mean to avoid desynchronization between the host's and device's counters. Device calculates up to 9 values ahead of its current counter to find the matching code (in total it calculates HOTP code for 10 subsequent counter positions). In case:
+    ///
+    /// - no code would match - the on-device counter will not be changed;
+    /// - incoming code parsing would fail - the on-device counter will not be changed;
+    /// - code would match, but with some counter's offset (up to 9) - the on-device counter will be set to matched code-generated HOTP counter and incremented by 1;
+    /// - code would match, and the code matches counter without offset - the counter will be incremented by 1.
+    ///
+    /// Device will stop verifying the HOTP codes in case, when the difference between the host and on-device counters will be greater or equal to 10.
+    ///
+    /// ```
```
[the remark this answers is not in the document]
Comment (szszszsz) [reply src/authenticator.rs]: Done

File: src/command.rs (hunk 2 above)
Comment (szszszsz) [reply src/command.rs:147]: Indeed. To be moved to a separate ticket. This specific implementation is just following general convention in this project.

Comment (sosthene-nitrokey) [reply src/command.rs:147]: I think it shouldn't be possible to trigger a panic with invalid input.

File: src/authenticator.rs (hunk 3 above)
Comment (szszszsz) [reply src/authenticator.rs]: Correct. Leaving at blinking at the moment. To be updated later.
- https://github.com/Nitrokey/trussed/issues/5

File: Cargo.toml [hunk 6]
```diff
@@ -20,14 +19,34 @@ hex-literal = "0.3"
...
-trussed = "0.1"
+trussed = "0.1.0"
+
+# ctaphid
+ctaphid-dispatch = { version = "0.1", optional = true }
+usbd-ctaphid = { git = "https://github.com/Nitrokey/nitrokey-3-firmware", optional = true }
+
+[dev-dependencies]
+log = { version = "0.4.14", default-features = false }
+clap = { version = "3.0.0", features = ["derive"] }
```
[the remark this answers is not in the document]
Comment (szszszsz) [reply Cargo.toml:30]: Removing these for now, as the example was not finished.

File: src/authenticator.rs (hunk 1 above)
Comment (szszszsz) [reply src/authenticator.rs]: Conversion removed entirely.

File: src/authenticator.rs (hunk 4 above)
Comment (szszszsz) [reply src/authenticator.rs]: Yes. Mentioned in https://github.com/Nitrokey/oath-authenticator/issues/4

File: src/authenticator.rs [hunk 7]
```diff
@@ -813,4 +911,48 @@ impl<T> hid::App for Authenticator<T>
...
+/// assert_eq!(convert_string_to_integer(b"123h"), Err(Status::ConditionsOfUseNotSatisfied));
+/// ```
+pub fn convert_string_to_integer(str: &[u8]) -> iso7816::Result<u32> {
+    if str.len() > 8 {
+        // Handle small codes only for now
+        return Err(Status::ConditionsOfUseNotSatisfied);
+    }
+    let mut number = 0u32;
+    for e in str.iter().rev().enumerate() {
+        let (i, digit) = e;
```
[the remark this answers is not in the document]
Comment (szszszsz) [reply src/authenticator.rs]: Removed entirely

**Pull request closed without merging**